device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/MultiplexingClient.java

```diff
@@ -0,0 +1,232 @@
+     * <p>
+     * Once closed, this client can be re-opened. It will preserve the previously registered device clients.
+     * <p>
+     * @throws IOException If any exception occurs while closing the connection.
+     */
+    public void close() throws IOException
+    {
+        log.info("Closing multiplexing client");
+        for (DeviceClient deviceClient : this.deviceClientList)
+        {
+            deviceClient.closeFileUpload();
+        }
+
+        if (this.deviceIO != null)
+        {
+            this.deviceIO.multiplexClose();
+        }
+
+        log.info("Successfully closed multiplexing client");
+    }
+
+    /**
+     * Add a device to this multiplexing client. This method may be called before or after opening the multiplexed
+     * connection, but will behave differently depending on when it was called.
+     * <p>
+     * Up to 1000 devices can be registered on a multiplexed AMQPS connection, and up to 500 devices can be registered on a
+     * multiplexed AMQPS_WS connection.
+     * <p>
+     * If the multiplexing client is already open, then this device client will automatically
+     * be opened, too. If the multiplexing client is not open yet, then this device client will not be opened until
+     * {@link MultiplexingClient#open()} is called.
+     * <p>
+     * If the multiplexed connection is already open, then this is an asynchronous operation. You can track the state of your
+     * device session using the {@link DeviceClient#registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback, Object)}.
+     * That callback will execute once your device session has successfully been added to the existing multiplexed connection and
+     * is ready to send telemetry.
+     * <p>
+     * Any proxy settings set to the provided device client will be overwritten by the proxy settings of this multiplexing client.
+     * <p>
+     * The registered device client instance must use the same transport protocol (AMQPS or AMQPS_WS) that this multiplexing client uses.
+     * <p>
+     * The registered device client may have its own retry policy and its own SAS token expiry time, separate from every other multiplexed device.
+     * <p>
+     * The registered device client must use symmetric key based authentication.
+     * <p>
+     * The registered device client must belong to the same IoT Hub as all previously registered device clients.
+     * <p>
+     * If this device client is already registered, then this method will throw a {@link IllegalStateException}
+     * <p>
+     * @param deviceClient The device client to associate with this multiplexing client.
+     */
+    public void registerDeviceClient(DeviceClient deviceClient)
+    {
+        Objects.requireNonNull(deviceClient);
+
+        for (DeviceClient currentClient : deviceClientList)
+        {
+            if (currentClient.getConfig().getDeviceId().equals(deviceClient.getConfig().getDeviceId()))
```

Comment:
should we throw in this case? can't we just be resilient to this?

Reply:
We'd need to at least log a warning about attempting to re-register an already registered client. I'm of the mindset that anyone who does that is probably doing it accidentally, and should be notified more strongly with this exception

Reply:
I don't disagree, but throwing is not just a warning. I don't feel too strongly about this, but I think if we can move on with the operation with no setbacks or issues, we should try doing that. this seems a fairly innocent operation to add an already existing device to multiplexing client.

Reply:
Fair enough. I'll make this just log a warning instead. I guess it is analogous to trying to open an already open device client, and we do allow that without throwing